Flow part expanded state configuration

hawkeye101 opened the pull request:
<!--- Please make sure that you're familiar with our contribution guidelines before submitting a pull request: https://docs.orchardcore.net/en/latest/contributing/. -->
Issue https://github.com/OrchardCMS/OrchardCore/issues/17503. I have added a toggleable setting in flow part and bag part settings regarding expansion, but need help on how to change this line:
`<div class="widget widget-editor card my-1@(Model.Inline ? " collapsed" : string.Empty)">
`

MikeAlhayek commented:
@hawkeye101 Thanks for working on this.

First the new property `AutoCollapseContainedItems` should be part of `FlowPartSettings` not `ContentPartSettings`. And there won't be a need for the `AutoCollapseContainedItems()` extension in your case.

> need help on how to change this line: <div class="widget widget-editor card my-1@(Model.Inline ? " collapsed" : string.Empty)"> 

In the `Content-Card-FlowPart.Edit.cshtml` view, the line https://github.com/OrchardCMS/OrchardCore/blob/22dd2447a9f69a8bfeb52baa2c2fa4dc58912020/src/OrchardCore.Modules/OrchardCore.Flows/Views/ContentCard-FlowPart.Edit.cshtml#L8 

will be changed to this:

```
    var contentTypeDefinition = await ContentDefinitionManager.GetTypeDefinitionAsync((string)Model.ContentItem.ContentType);
    var contentTypeDisplayText = contentTypeDefinition.DisplayName;
    var flowPartSettings = contentTypeDefinition.GetSettings<FlowPartSettings>();
```

Then the HTML will be changed to this

```
<div class="widget widget-editor card my-1@(Model.Inline || flowPartSettings.AutoCollapseContainedItems ? " collapsed" : string.Empty)">
```

sebastienros commented:
I see it's still missing the part that actually collapses the part.

Naming: Why "auto"? What is automatic? How would that be different than just "collapsed"? Is that a default state when the editor is displayed? If so then we can find a better name (`CollapseContentItemEditors`, `CollapseContentItems`, `CollapseEditors`).

Since the current value is expanded (not collapsed) the value will be false by default (good right now).
Update the hint to remove "auto" also

hawkeye101 commented:
@MikeAlhayek Thanks for the detailed response. What about bagPartSettings? This is the current code:
`    var collapsed = !Model.Inline;

    if (!ViewContext.ModelState.IsValid && ViewContext.ModelState.Keys.Any(key => key.StartsWith(Model.PrefixValue, StringComparison.OrdinalIgnoreCase)))
    {
        collapsed = false;
    }
<div class="widget widget-editor card@(collapsed ? " collapsed" : string.Empty)">
`
I changed it to 
`    var collapsed = !Model.Inline || bagPartSettings.AutoCollapseContainedItems;
`
Also, since we removed it from ContentPartSettings, there is no migration, How to add it there?

hawkeye101 commented:
@sebastienros I'll remove the 'Auto' but do we need a migration to change the default value of the setting?

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Models/BagPartSettings.cs:
```diff
@@ -7,4 +7,5 @@ public class BagPartSettings
     public string[] ContainedStereotypes { get; set; } = [];
 
     public string DisplayType { get; set; }
+    public bool CollapseContainedItems { get; set; }
```
```suggestion

    public bool CollapseContainedItems { get; set; }
```

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Models/FlowPartSettings.cs:
```diff
@@ -3,4 +3,5 @@ namespace OrchardCore.Flows.Models;
 public class FlowPartSettings
 {
     public string[] ContainedContentTypes { get; set; } = [];
+    public bool CollapseContainedItems { get; set; }
```
```suggestion

    public bool CollapseContainedItems { get; set; }
```

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Settings/FlowPartSettingsDisplayDriver.cs:
```diff
@@ -24,6 +24,7 @@ public override IDisplayResult Edit(ContentTypePartDefinition contentTypePartDef
             model.FlowPartSettings = contentTypePartDefinition.GetSettings<FlowPartSettings>();
             model.ContainedContentTypes = model.FlowPartSettings.ContainedContentTypes;
             model.ContentTypes = [];
+            model.CollapseContainedItems = model.FlowPartSettings.CollapseContainedItems;
```
Move it up one line so its listed under the other flow part settings

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Views/BagPartSettings.Edit.cshtml:
```diff
@@ -39,6 +39,11 @@
 </div>
+<div class="form-check">
+    <input type="checkbox" class="form-check-input" asp-for="CollapseContainedItems">
+    <label class="form-check-label" asp-for="CollapseContainedItems">@T["Auto Collapse"]</label>
```
```suggestion
    <label class="form-check-label" asp-for="CollapseContainedItems">@T["Auto Collapse Contained Items"]</label>
```

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Views/BagPartSettings.Edit.cshtml:
```diff
@@ -39,6 +39,11 @@
+<div class="form-check">
+    <input type="checkbox" class="form-check-input" asp-for="CollapseContainedItems">
+    <label class="form-check-label" asp-for="CollapseContainedItems">@T["Auto Collapse"]</label>
+    <span class="hint">@T["Check to auto collapse contained items."]</span>
```
```suggestion
    <span class="hint">@T["When checked, the contained items will be collapsed when the edit form is loaded."]</span>
```

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.ContentTypes/Views/ContentPartSettings.Edit.cshtml:
```diff
@@ -30,4 +30,4 @@
     <label class="form-check-label" asp-for="Reusable">@T["Reusable"]</label>
     <span class="hint">@T["Check to allow this part to be attached multiple times."]</span>
-</div>
+</div>
```
```suggestion
</div>

```

MikeAlhayek commented on src/OrchardCore.Modules/OrchardCore.Flows/Views/ContentCard-BagPart.Edit.cshtml:
```diff
@@ -4,8 +4,10 @@
+    var bagPartSettings = contentTypeDefinition.GetSettings<OrchardCore.Flows.Models.BagPartSettings>();
+    var collapsed = !Model.Inline || bagPartSettings.CollapseContainedItems;
 
     if (!ViewContext.ModelState.IsValid && ViewContext.ModelState.Keys.Any(key => key.StartsWith(Model.PrefixValue, StringComparison.OrdinalIgnoreCase)))
```
Can we add this check to the flow part too?

MikeAlhayek reviewed (changes requested):
Thanks for your effort on this PR.

We need to keep the code in flow-part and bag-part consistent by default. I don't recall the existing behavior. If the current behavior is different we should add a migration to ensure existing tenants behavior doesn't change. 

The content should be visible 
1. Model.Inline is true
2. Or, there is an error in the contained item. (This is something we do in the bag part but should also be added to flow part)
3. Or, if the new setting is true.

hawkeye101 commented:
I made the suggested changes, but there is an issue with the runtime in the setup so I'm not able to login and test.

MikeAlhayek commented:
@hawkeye101 I made some changes to your PR. I’ll be making few more changes to it in the morning.
